### PR TITLE
Blazor WebAssembly caching fixes

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This PR addresses a few potential issues:

1. It now aims to handle *all* scenarios where the browser will not let us cache resources. Previously we were assuming that the CacheStorage APIs would always be available. However that will not be the case if (1) you're running on an insecure origin, or (2) you hit the Chromium bug where it won't give you a valid cache if the `--user-data-dir` path is too long, or (3) the `caches.open` promise rejects for any other reason that I haven't anticipated. All these cases are now handled in a common way, in which we proceed without using CacheStorage.
2. It auto-adds a cache-busting parameter when requesting `dotnet.js`, as well as enforcing SRI checks like we do for other subresources. This is the outcome from [this discussion](https://github.com/dotnet/aspnetcore/issues/19199). It means that developers can now safely deploy to servers that can't be configured to return `cache-control: no-cache` on responses, such as GitHub pages. The only extra manual step needed when deploying to such a server is manually adding a cache-busting parameter in your `<script src="_framework/blazor.webassembly.js">` tag in `index.html`, and then we can be sure that all framework-controlled subresources are fetched freshly regardless of server cache headers.

This PR also slightly cleans up the organization of `blazor.boot.json`. Previously we categorized by file extension, assuming that's enough to know the purpose of the file. With this PR, we let the MSBuild targets choose an arbitrary resource type based on what it knows about why the file is being included.

The diff in the TypeScript looks quite noisy, but very little functional behavior has actually changed. The noise is mostly due to renames and extract-method refactoring.

### CODE REVIEW

**Review by javiercn — COMMENTED** [review]
The changes look great.

I imagine there's no way we can cover this through functional tests. Do you think we can have at least a CTI scenario to cover deploying to an HTTP only site? I'm not even sure where you can host such a thing these days, but I'm sure its doable.

**Comment by SteveSandersonMS** [comment]
@javiercn Yes, CTI tests will be the only reasonable way to cover the insecure-origin scenario. I verified it locally by adding an entry to my OS `Hosts` file mapping to 127.0.0.1.

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 1]
```diff
@@ -112,15 +112,29 @@ export const monoPlatform: Platform = {
...
     throw new Error('This browser does not support WebAssembly.');
   }
 
+  // The dotnet.*.js file has a version or hash in its name as a form of cache-busting. This is needed
+  // because it's the only part of the loading process that can't use cache:'no-cache' (because it's
+  // not a 'fetch') and isn't controllable by the developer (so they can't put in their own cache-busting
+  // querystring). So, to find out the exact URL we have to search the boot manifest.
+  const dotnetJsResourceName = Object
+    .keys(resourceLoader.bootConfig.resources.runtime)
+    .filter(n => n.startsWith('dotnet.') && n.endsWith('.js'))[0];
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts]: Can this conflicts with a user resource (or is the `runtime` section special)?

File: src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts [hunk 2]
```diff
@@ -12,19 +12,13 @@ export class WebAssemblyResourceLoader {
...
-    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
-    const cacheName = `blazor-resources-${relativeBaseHref}`;
-    return new WebAssemblyResourceLoader(
-      await bootConfigResponse.json(),
-      await caches.open(cacheName));
+    return new WebAssemblyResourceLoader(bootConfig, cache);
   }
 
-  constructor (public readonly bootConfig: BootJsonData, private cache: Cache)
+  constructor (public readonly bootConfig: BootJsonData, private cacheIfUsed: Cache | null)
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts]: I kinda prefer calling this `cache` and relying on the nullability.

**Review by pranavkm — APPROVED** [review]

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts]: Everything in `blazor.boot.json` comes from the framework - there are no user-supplied files there.

File: src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts]: I've kept the scope of `cacheIfUsed` as tight as possible - it explicitly creates a separate variable called `cache` when nonnull. So I kind of like having this extra clarity here that a null value is not a random error to be worked around via saying `this.cache!`, but is an explicit scenario we need to support for good reasons.

If you feel strongly let me know, but for now I'll proceed as-is.

**Comment by szalapski** [comment]
Is there a suggestion of how to go about cache-busting .css files linked in index.html?  I'd rather not make it a .cshtml.

**Comment by SteveSandersonMS** [comment]
@szalapski Generally as long as your web server returns the correct HTTP cache control headers, there will be no need to do anything extra to invalidate caches. For example if your responses are served with Etag values, then browsers will send the `If-Modified` headers so that the server can either return updated content or a 304 Not Modified response. With HTTP/2 and later, there's not much of a per-HTTP-request overhead since this will be pipelined over an existing connection.

If the issue is that you can't configure your server to return correct HTTP headers, then you probably will have to vary the URLs when the content changes. If you can't change the host page to be server-rendered (e.g., `.cshtml`) then you will probably have to do that manually or via some custom build process.

**Pull request merged**